File: rust/src/models/coupon.rs
// this file is @generated
use serde::{Deserialize, Serialize};

use super::{coupon_discount::CouponDiscount, coupon_id::CouponId, plan_id::PlanId};

#[derive(Clone, Debug, Default, PartialEq, Deserialize, Serialize)]
pub struct Coupon {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub archived_at: Option<String>,

    pub code: String,

    pub created_at: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,

    pub disabled: bool,

    pub discount: CouponDiscount,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub expires_at: Option<String>,

    pub id: CouponId,

    pub plan_ids: Vec<PlanId>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub recurring_value: Option<i32>,

    pub redemption_count: i32,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub redemption_limit: Option<i32>,

    pub reusable: bool,
}

impl Coupon {
    pub fn new(
        code: String,
        created_at: String,
        disabled: bool,
        discount: CouponDiscount,
        id: CouponId,
        plan_ids: Vec<PlanId>,
        redemption_count: i32,
        reusable: bool,
    ) -> Self {
        Self {
            archived_at: None,
            code,
            created_at,
            description: None,
            disabled,
            discount,
            expires_at: None,
            id,
            plan_ids,
            recurring_value: None,
            redemption_count,
            redemption_limit: None,
            reusable,
        }
    }
}

File: typescript/src/models/coupon.ts
// this file is @generated
import { parseDateTime } from "../datetime";
import { type CouponDiscount, CouponDiscountSerializer } from "./couponDiscount";
import { type CouponId, CouponIdSerializer } from "./couponId";
import { type PlanId, PlanIdSerializer } from "./planId";

export interface Coupon {
  archivedAt?: Date | null;

  code: string;

  createdAt: Date;

  description?: string | null;

  disabled: boolean;

  discount: CouponDiscount;

  expiresAt?: Date | null;

  id: CouponId;

  planIds: PlanId[];

  recurringValue?: number | null;

  redemptionCount: number;

  redemptionLimit?: number | null;

  reusable: boolean;
}

export const CouponSerializer = {
  _fromJsonObject(object: any): Coupon {
    return {
      archivedAt:
        object["archived_at"] != null ? parseDateTime(object["archived_at"]) : undefined,
      code: object["code"],
      createdAt: parseDateTime(object["created_at"]),
      description: object["description"],
      disabled: object["disabled"],
      discount: CouponDiscountSerializer._fromJsonObject(object["discount"]),
      expiresAt:
        object["expires_at"] != null ? parseDateTime(object["expires_at"]) : undefined,
      id: CouponIdSerializer._fromJsonObject(object["id"]),
      planIds: object["plan_ids"].map((item: any) =>
        PlanIdSerializer._fromJsonObject(item)
      ),
      recurringValue: object["recurring_value"],
      redemptionCount: object["redemption_count"],
      redemptionLimit: object["redemption_limit"],
      reusable: object["reusable"],
    };
  },

  _toJsonObject(self: Coupon): any {
    return {
      archived_at: self.archivedAt,
      code: self.code,
      created_at: self.createdAt,
      description: self.description,
      disabled: self.disabled,
      discount: CouponDiscountSerializer._toJsonObject(self.discount),
      expires_at: self.expiresAt,
      id: CouponIdSerializer._toJsonObject(self.id),
      plan_ids: self.planIds.map((item: any) => PlanIdSerializer._toJsonObject(item)),
      recurring_value: self.recurringValue,
      redemption_count: self.redemptionCount,
      redemption_limit: self.redemptionLimit,
      reusable: self.reusable,
    };
  },
};
